templates/README.md

```diff
@@ -0,0 +1,221 @@
+
+Also, there might be multiple different OSes, flavors, sizes mentioned by any single template file if they share the same configuration.
+
+## User Experience
+
+The flow should be the same as usual with regards to openshift templates. The user either takes a local template and generates a VM object.
+
+Or the user gets a template from kubernetes and invokes the proper REST command on top of it to keep the VM linked to the template or edit it in accordance to the template.
+
+The UI can operate on top of template parameters and/or the Template.VM.spec DOM directly.
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    description: "Basic windows 10 template"
+    
+    # This set of annotations is meant as a hint only, it is
+    # not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded here informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+    defaults.template.cnv.io/disk: default-disk
+    defaults.template.cnv.io/volume: default-volume
+    defaults.template.cnv.io/network: default-network
+
+    # Replacements useful for finishing the transformation
+    # This is an extension we could use for specifying common
+    # fields for all disks/nics. Requires extra implementation
+    # that will do the find and replace using a syntax documented
+    # here: https://kubernetes.io/docs/reference/kubectl/jsonpath/
+    template.cnv.io/replace:
+    - path: "/VM/disks/*/type"
+      value: virtio
+
+    # Another extension useful for CNV aware tooling that allows
+    # expressing additional validation rules for this template
+    template.cnv.io/validate:
+    - path: "/VM/spec/domain/guest/memory"
+      min: 3
+      max: 4
+
+    # Another extension for hinting at which elements should be
+    # considered editable
+    template.cnv.io/editable:
+      - "/VM/spec/domain/guest/memory"
+      - "/VM/spec/domain/guest/memory"
+
+  labels:
+    # The UI can show all possible template.cnv.io/* values in a nice way
+    # and let the user filter down the available templates to the one
+    # the user actually wants:
+    # A single selected template only means no conflicts and no smart
+    # merging code. This has to be done using labels to allow efficient
+    # searching.
+    # The format has the following meaning:
+    # {os,flavor,size}.template.cnv.io/{value}: {version} or just true
+    os.template.cnv.io/windows2k12r2: 201809 
+    os.template.cnv.io/windows8: 201809 
+    os.template.cnv.io/windows7: 201809 
+    flavor.template.cnv.io/minimal: 201809
+    # io.cnv.vm.size not specified means all?
```

Comment:
We should require all the main parts to be set, os, size, workload.

If we want to say "ALL", then we might want to use such a special keyword i.e. `size: ALL`

The names "windows8" should rather be the libosinfo identifier whhich provides unique ids for all major osses. I also have the mipression that the os should be a value, rather than a key.

All in all I'm not yet sure about how the fields shoudl excatly look.

Reply:
These are annotations and here we've got the complete freedom. Take a look at these examples: https://docs.okd.io/latest/dev_guide/templates.html#writing-description